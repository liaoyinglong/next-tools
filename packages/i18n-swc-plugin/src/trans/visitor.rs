use swc_core::ecma::ast::{JSXAttrName, JSXAttrOrSpread, JSXElement};
use swc_core::ecma::atoms::JsWord;
use swc_core::ecma::visit::VisitMut;
use tracing::error;

use crate::shared::Normalizer;

static TRANS_COMMENT_NAME: &str = "Trans";
pub struct TransVisitor;

impl VisitMut for TransVisitor {
    // Implement necessary visit_mut_* methods for actual custom transform.
    // A comprehensive list of possible visitor methods can be found here:
    // https://rustdoc.swc.rs/swc_ecma_visit/trait.VisitMut.html
    fn visit_mut_jsx_element(&mut self, n: &mut JSXElement) {
        let mut work = || -> Option<()> {
            let mut normalizer = Normalizer::new();
            let is_trans_component = Normalizer::get_jsx_element_name(n)? == TRANS_COMMENT_NAME;
            if !is_trans_component {
                return None;
            }
            if n.children.len() == 0 {
                error!("Trans component must have at least one child");
                return None;
            }
            let attrs = &mut n.opening.attrs;

            normalizer.jsx_children_work(n.children.clone());
            n.children.clear();

            let exist_id_prop = attrs.iter().any(|attr| {
                if let JSXAttrOrSpread::JSXAttr(attr) = attr {
                    if let JSXAttrName::Ident(ident) = &attr.name {
                        return ident.sym == JsWord::from("id");
                    }
                }
                false
            });
            attrs.extend_from_slice(&normalizer.to_jsx_attr(exist_id_prop));

            None
        };

        work();
    }
}
