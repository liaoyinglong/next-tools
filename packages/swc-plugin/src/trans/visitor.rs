use swc_core::ecma::ast::{JSXAttrName, JSXAttrOrSpread, JSXElement, JSXElementName};
use swc_core::ecma::atoms::JsWord;
use swc_core::ecma::visit::VisitMut;
use swc_core::ecma::visit::VisitMutWith;
use tracing::error;

use crate::shared::Normalizer;

static TRANS_COMMENT_NAME: Option<&str> = Some("Trans");
pub struct TransVisitor;

impl TransVisitor {
    fn get_jsx_element_name(element: &mut JSXElement) -> Option<&str> {
        if let JSXElementName::Ident(ident) = &element.opening.name {
            return Some(&ident.sym);
        }
        None
    }
    fn is_trans_component(element: &mut JSXElement) -> bool {
        Self::get_jsx_element_name(element) == TRANS_COMMENT_NAME
    }
}

impl VisitMut for TransVisitor {
    // Implement necessary visit_mut_* methods for actual custom transform.
    // A comprehensive list of possible visitor methods can be found here:
    // https://rustdoc.swc.rs/swc_ecma_visit/trait.VisitMut.html
    fn visit_mut_jsx_element(&mut self, n: &mut JSXElement) {
        n.visit_mut_children_with(self);

        if !Self::is_trans_component(n) {
            return;
        }
        if n.children.len() == 0 {
            error!("Trans component must have at least one child");
            return;
        }
        let attrs = &mut n.opening.attrs;

        let mut normalizer = Normalizer::new();
        normalizer.jsx_children_work(n.children.clone());
        n.children.clear();

        let exist_id_prop = attrs.iter().any(|attr| {
            if let JSXAttrOrSpread::JSXAttr(attr) = attr {
                if let JSXAttrName::Ident(ident) = &attr.name {
                    return ident.sym == JsWord::from("id");
                }
            }
            false
        });
        attrs.extend_from_slice(&normalizer.to_jsx_attr(exist_id_prop));
    }
}
