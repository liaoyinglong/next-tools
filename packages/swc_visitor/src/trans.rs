use crate::shared::Normalizer;
use swc_core::ecma::ast::{JSXAttrName, JSXAttrOrSpread, JSXElement, JSXElementName};
use swc_core::ecma::atoms::JsWord;
use swc_core::ecma::visit::VisitMutWith;
use swc_core::ecma::visit::{noop_visit_mut_type, VisitMut};
use tracing::error;

pub struct TransVisitor;

impl TransVisitor {
    fn is_trans_component(element: &mut JSXElement) -> bool {
        if let JSXElementName::Ident(ident) = &element.opening.name {
            if ident.sym == JsWord::from("Trans") {
                // re-create new ident to make same syntaxContext
                // element.opening.name = JSXElementName::Ident(quote_ident!(ident.sym.to_string()));
                return true;
            }
        }
        false
    }
}

impl VisitMut for TransVisitor {
    // Implement necessary visit_mut_* methods for actual custom transform.
    // A comprehensive list of possible visitor methods can be found here:
    // https://rustdoc.swc.rs/swc_ecma_visit/trait.VisitMut.html
    noop_visit_mut_type!();

    fn visit_mut_jsx_element(&mut self, n: &mut JSXElement) {
        n.visit_mut_children_with(self);

        if !Self::is_trans_component(n) {
            return;
        }
        if n.children.len() == 0 {
            error!("Trans component must have at least one child");
            return;
        }

        let attrs = &mut n.opening.attrs;

        let mut normalizer = Normalizer::new();
        normalizer.jsx_children_work(n.children.clone());
        n.children.clear();

        let exist_id_prop = attrs.iter().any(|attr| {
            if let JSXAttrOrSpread::JSXAttr(attr) = attr {
                if let JSXAttrName::Ident(ident) = &attr.name {
                    return ident.sym == JsWord::from("id");
                }
            }
            false
        });
        attrs.extend_from_slice(&normalizer.to_jsx_attr(exist_id_prop));
    }
}
