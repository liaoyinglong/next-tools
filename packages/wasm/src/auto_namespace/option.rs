use std::sync::Arc;
use swc_core::common::errors::HANDLER;
use swc_core::common::sync::Lrc;
use swc_core::common::SourceMap;
use swc_core::ecma::ast::{
    CallExpr, Expr, ExprOrSpread, JSXAttr, JSXAttrName, JSXAttrOrSpread, JSXAttrValue,
    JSXElementName, JSXExpr, JSXOpeningElement, Lit, ObjectLit, Str,
};
use swc_core::ecma::atoms::JsWord;
use swc_core::ecma::utils::quote_ident;
use swc_core::ecma::visit::VisitMutWith;
use swc_core::ecma::visit::{noop_visit_mut_type, VisitMut};

#[derive(Clone)]
pub struct AutoNamespaceOption {
    /// 源代码
    pub source: String,
    /// 翻译key的前缀
    pub namespace: String,
    /// 用于分割 namespace 和 key 的字符串
    pub separator: String,

    /// Trans component name
    /// default is Trans
    pub trans_component: String,
    // Trans component  ia attr name
    // default is id
    pub id_attr: String,

    // 默认的翻译函数 t
    pub t_fn: String,
}

impl AutoNamespaceOption {
    // 添加 namespace
    fn add_namespace(&mut self, str: String) -> Option<String> {
        // 已经带有 namespace 前缀的不处理
        let x: Vec<_> = str.split(self.separator.as_str()).collect();
        if x.len() > 1 {
            return None;
        }

        Some(format!("{}{}{}", self.namespace, self.separator, str))
    }

    // Trans 组件 处理
    fn get_jsx_attr_or_spread(&mut self, str: String) -> Option<JSXAttrOrSpread> {
        let v = self.add_namespace(str)?;
        let v = JsWord::from(v);
        Some(JSXAttrOrSpread::JSXAttr(JSXAttr {
            span: Default::default(),
            name: JSXAttrName::Ident(quote_ident!(self.id_attr.clone())),
            value: Some(JSXAttrValue::Lit(Lit::Str(Str {
                span: Default::default(),
                value: v,
                raw: None,
            }))),
        }))
    }

    fn append_to_lit(&mut self, lit: Lit) -> Option<JSXAttrOrSpread> {
        match lit {
            Lit::Str(str) => self.get_jsx_attr_or_spread(str.value.to_string()),
            Lit::Num(str) => self.get_jsx_attr_or_spread(str.value.to_string()),
            _ => None,
        }
    }

    fn find_attr_name(attr: JSXAttrOrSpread) -> Option<String> {
        if let JSXAttrOrSpread::JSXAttr(attr) = attr {
            if let JSXAttrName::Ident(id) = attr.name.clone() {
                return Some(id.sym.to_string());
            }
        };
        None
    }

    // t fn 处理
}

impl Default for AutoNamespaceOption {
    fn default() -> Self {
        AutoNamespaceOption {
            source: "".to_string(),
            namespace: "".to_string(),
            separator: ".".to_string(),
            trans_component: "Trans".to_string(),
            id_attr: "id".to_string(),
            t_fn: "t".to_string(),
        }
    }
}

impl VisitMut for AutoNamespaceOption {
    // Implement necessary visit_mut_* methods for actual custom transform.
    // A comprehensive list of possible visitor methods can be found here:
    // https://rustdoc.swc.rs/swc_ecma_visit/trait.VisitMut.html
    noop_visit_mut_type!();

    /// 对应 t 方法调用的时候有以下：
    /// ```js
    /// t({ id: "Refresh inbox", message: "Refresh inbox" });
    /// t({ id: "Refresh inbox" }, { name: "name" });
    /// t({ id: "Refresh inbox", values: { name: "name" }});
    /// t("Refresh inbox", { name: "name" });
    /// ```
    fn visit_mut_call_expr(&mut self, n: &mut CallExpr) {
        n.visit_mut_children_with(self);
        let new_args = vec![];

        let mut work = || -> Option<()> {
            let ident = n.callee.as_expr()?.as_ident()?;
            if ident.sym.to_string() != self.t_fn {
                return None;
            }
            let arg = &n.args;

            let first_arg = arg.first()?;
            let mut id = "".to_string();

            match *first_arg.expr.clone() {
                Expr::Lit(lit) => {
                    let v = self.add_namespace(lit.value.to_string())?;

                    let id_arg = ExprOrSpread {
                        spread: first_arg.spread.clone(),
                        expr: Box::new(Expr::Lit(Lit::Str(Str {
                            span: Default::default(),
                            value: JsWord::from(v),
                            raw: None,
                        }))),
                    };
                    // n.args.0 = id_arg;
                    None
                }
                // Expr::Tpl(tpl) => {
                //     if tpl.exprs.is_empty() {
                //         id = tpl.quasis.get(0)?.raw.to_string();
                //     }
                // }
                // TODO：目前代码里没有人用 t({ id: "Refresh inbox", message: "Refresh inbox" }) 这种写法
                // Expr::Object(obj) => {
                //     id = self.pick_object_value(obj.clone(), &self.config.id.clone())?;
                //     default_msg = self.pick_object_value(obj, &self.config.message.clone())?;
                // }
                _ => None,
            }
        };
        work();
    }

    // case: <Trans id="msg.refresh" message="Refresh inbox" />
    fn visit_mut_jsx_opening_element(&mut self, n: &mut JSXOpeningElement) {
        n.visit_mut_children_with(self);
        let el_name = match &n.name {
            JSXElementName::Ident(ident) => ident.sym.to_string(),
            _ => "".to_string(),
        };
        if el_name != self.trans_component {
            return;
        }

        let mut new_attrs: Vec<JSXAttrOrSpread> = vec![];

        n.attrs.iter().for_each(|attr| {
            let mut f = || -> Option<()> {
                let attr_name = Self::find_attr_name(attr.clone())?;
                if attr_name != self.id_attr {
                    new_attrs.push(attr.clone());
                    return None;
                }
                // 处理 对应的 attr
                let mut get_new_attr = || -> Option<JSXAttrOrSpread> {
                    if let JSXAttrOrSpread::JSXAttr(jsx_attr) = attr {
                        let v = jsx_attr.value.clone()?;
                        return match v {
                            JSXAttrValue::Lit(lit) => self.append_to_lit(lit),
                            JSXAttrValue::JSXExprContainer(container) => match container.expr {
                                JSXExpr::Expr(expr) => match *expr {
                                    Expr::Lit(lit) => self.append_to_lit(lit),
                                    Expr::Tpl(tpl) => {
                                        if tpl.exprs.is_empty() {
                                            let v = tpl
                                                .quasis
                                                .get(0)
                                                .map(|item| item.raw.to_string())?;

                                            return self.get_jsx_attr_or_spread(v);
                                        }
                                        // case <Trans id={`msg_${name}`} />, 这种不支持
                                        None
                                    }
                                    _ => None,
                                },
                                _ => None,
                            },
                            _ => None,
                        };
                    }
                    None
                };
                let new_lit = get_new_attr();
                match new_lit {
                    Some(new_attr) => new_attrs.push(new_attr),
                    _ => new_attrs.push(attr.clone()),
                };

                None
            };
            f();
        });

        n.attrs.clear();
        n.attrs.extend_from_slice(&*new_attrs)
    }
}
