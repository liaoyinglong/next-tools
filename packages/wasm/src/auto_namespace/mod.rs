pub mod option;
use s_swc_visitor::trans::TransVisitor;

use crate::auto_namespace::option::AutoNamespaceOption;
use anyhow::Error;
use swc_core::common::comments::SingleThreadedComments;
use swc_core::common::sync::Lrc;
use swc_core::common::{FileName, SourceMap};
use swc_core::ecma::ast::EsVersion::EsNext;
use swc_core::ecma::parser::lexer::Lexer;
use swc_core::ecma::parser::{Parser, StringInput, Syntax, TsConfig};
use swc_core::ecma::visit::VisitMutWith;
use swc_ecma_codegen::text_writer::JsWriter;
use swc_ecma_codegen::Emitter;

/// 自动给 t 、 Trans 调用加上 namespace 前缀
pub fn auto_namespace(mut opts: AutoNamespaceOption) -> Result<String, Error> {
    let cm = Lrc::<SourceMap>::default();
    let source_file = cm.new_source_file(FileName::Anon, opts.source.clone());
    let syntax = Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    });
    let comments = SingleThreadedComments::default();

    let lexer = Lexer::new(
        syntax,
        EsNext,
        StringInput::from(&*source_file),
        Some(&comments),
    );
    let mut parser = Parser::new_from(lexer);
    let mut module = parser
        .parse_typescript_module()
        .expect("Failed to parse file");

    module.visit_mut_with(&mut TransVisitor {});
    module.visit_mut_with(&mut opts);

    // 生成代码
    let code = {
        if opts.source_has_changed {
            let mut buf = vec![];
            let mut emitter = Emitter {
                cfg: Default::default(),
                cm: cm.clone(),
                comments: Some(&comments),
                wr: JsWriter::new(cm, "\n", &mut buf, None),
            };
            emitter.emit_module(&module).unwrap();
            String::from_utf8(buf).unwrap()
        } else {
            "".to_string()
        }
    };
    Ok(code)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run_test(source: &str, expected: &str) {
        let code = auto_namespace(AutoNamespaceOption {
            source: source.to_string(),
            namespace: "menu".to_string(),

            ..Default::default()
        })
        .unwrap();
        let code = code.trim_end_matches("\n");

        assert_eq!(code, expected);
    }

    #[test]
    fn t_fn_simple() {
        // 模版字符串
        run_test("t`msg3`;", "t`menu.msg3`;");
        run_test(
            "t(`msg4`, { count: 1 });",
            "t(`menu.msg4`, {\n    count: 1\n});",
        );

        // 模版字符串中有变量
        run_test("t`hello ${name}`;", "t`menu.hello ${name}`;");

        // 不需要转换
        run_test("t('common.msg4', { count: 1 });", "");
        run_test("t`common.msg5`;", "");
        run_test("t`common.msg5 ${name}`;", "");

        // 不支持的转换
        run_test("t(id);", "");
        run_test("t(id, { count: 1 });", "");
    }

    #[test]
    fn trans_simple() {
        run_test("<Trans id='msg1' />;", r#"<Trans id="menu.msg1"/>;"#);
        run_test(
            "<Trans id='msg1'>children</Trans>;",
            r#"<Trans id="menu.msg1" message="children"></Trans>;"#,
        );
        run_test("<Trans id={'msg1'} />;", r#"<Trans id="menu.msg1"/>;"#);

        // Transform macro
        run_test("<Trans>hello</Trans>", "<Trans id=\"menu.hello\"></Trans>;");
        run_test(
            "<Trans>hello {name}</Trans>",
            "<Trans id=\"menu.hello {name}\" values={{\n    name: name\n}}></Trans>;",
        );
        run_test(
            "<Trans> hello <span>{name}</span> </Trans>",
            "<Trans id=\"menu.hello <0>{name}</0>\" values={{\n    name: name\n}} components={{\n    0: <span>{name}</span>\n}}></Trans>;"
        );

        // 不支持的转换
        run_test("<Trans id={`${id}`} />;", "")
    }

    #[test]
    fn integrate() {
        run_test(
            r###"
        const obj =  { msg: t`name` };
        function fn() {
            return t`name`;
        };

        function App(){

            return <div>
                <Trans id='msg1'>children</Trans>
            </div>
        }

        "###,
            "const obj = {\n    msg: t`menu.name`\n};\nfunction fn() {\n    return t`menu.name`;\n}\n;\nfunction App() {\n    return <div>\n\n                <Trans id=\"menu.msg1\">children</Trans>\n\n            </div>;\n}",
        );
    }
}
