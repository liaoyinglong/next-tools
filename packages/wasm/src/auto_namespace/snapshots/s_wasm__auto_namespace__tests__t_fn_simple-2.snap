---
source: packages/wasm/src/auto_namespace/mod.rs
expression: "run_test(\"t(`msg4`, { count: 1 })\")"
---
"t(`menu.msg4`, {\n    count: 1\n});"
